File: py/day02.py
# %%
def part1(course):
    x, z = 0, 0
    for line in course:
        direction, step = line.split(' ')
        step = int(step)
        if direction == "forward":
            x += step
        elif direction == "down":
            z += step
        elif direction == "up":
            z -= step
    return x, z

# %%
def part2(course):
    x = z = a = 0
    for line in course:
        direction, step = line.split(' ')
        step = int(step)
        if direction == "down":
            a += step
        elif direction == "up":
            a -= step
        elif direction == "forward":
            x += step
            z += a * step
    return x, z

# %%
if __name__ == "__main__":
    with open("../data/day02.txt", 'r') as f:
        course = f.read().splitlines()

    x1, z1 = part1(course)
    print("Part 1 —", x1 * z1)
    x2, z2 = part2(course)
    print("Part 2 —", x2 * z2)

# %%

File: rs/day02.rs

fn part1(course: &Vec<(&str, usize)>) -> usize {
    let (mut h, mut d) = (0, 0);
    for (direction, x) in course {
        match *direction {
            "forward" => h += x,
            "down" => d += x,
            "up" => d -= x,
            _ => (),
        }
    }
    h * d
}

fn part2(course: &Vec<(&str, usize)>) -> usize {
    let (mut h, mut d, mut aim) = (0, 0, 0);
    for (direction, x) in course {
        match *direction {
            "forward" => {h += x; d += aim * x},
            "down" => aim += x,
            "up" => aim -= x,
            _ => (),
        }
    }
    h * d
}

fn main() {
    let course: Vec<(&str, usize)> = include_str!("input.txt")
        .lines()
        .map(|line| line.split_once(" ").unwrap())
        .map(|line| (line.0, line.1.parse().unwrap()))
        .collect();
    
    println!("Part 1 — {:?}", part1(&course));
    println!("Part 2 — {:?}", part2(&course));

}
